File: src/providers/zenvia/state.ts
import { randomUUID } from "node:crypto";
import { getDb } from "../../shared/storage/db.js";
import type { CreateSmsMessageInput, CreateSubscriptionInput } from "./schemas.js";

export interface ZenviaMessage {
  id: string;
  externalId: string | null;
  channel: string;
  direction: "IN" | "OUT";
  from: string;
  to: string;
  contents: unknown[];
  status: string;
  createdAt: string;
}

interface MessageRow {
  id: string;
  external_id: string | null;
  channel: string;
  direction: string;
  sender: string;
  recipient: string;
  contents_json: string;
  status: string;
  created_at: string;
}

function toMessage(row: MessageRow): ZenviaMessage {
  return {
    id: row.id,
    externalId: row.external_id,
    channel: row.channel,
    direction: row.direction as "IN" | "OUT",
    from: row.sender,
    to: row.recipient,
    contents: JSON.parse(row.contents_json),
    status: row.status,
    createdAt: row.created_at,
  };
}

export function createMessage(channel: string, input: CreateSmsMessageInput): ZenviaMessage {
  const db = getDb();
  const id = randomUUID();
  db.prepare(
    `INSERT INTO zenvia_messages (id, external_id, channel, direction, sender, recipient, contents_json, status)
     VALUES (?, ?, ?, 'OUT', ?, ?, ?, 'ACCEPTED')`,
  ).run(id, input.externalId ?? null, channel, input.from, input.to, JSON.stringify(input.contents));

  const row = db.prepare("SELECT * FROM zenvia_messages WHERE id = ?").get(id) as MessageRow;
  return toMessage(row);
}

export function listMessages(): ZenviaMessage[] {
  const rows = getDb().prepare("SELECT * FROM zenvia_messages ORDER BY created_at DESC").all() as MessageRow[];
  return rows.map(toMessage);
}

export function getMessage(id: string): ZenviaMessage | undefined {
  const row = getDb().prepare("SELECT * FROM zenvia_messages WHERE id = ?").get(id) as MessageRow | undefined;
  return row ? toMessage(row) : undefined;
}

export function updateMessageStatus(id: string, status: string): void {
  getDb().prepare("UPDATE zenvia_messages SET status = ? WHERE id = ?").run(status, id);
}

export interface ZenviaSubscription {
  id: string;
  eventType: string;
  webhookUrl: string;
  webhookHeaders: Record<string, string> | null;
  criteriaChannel: string;
  criteriaDirection: string | null;
  status: string;
  createdAt: string;
  updatedAt: string;
}

interface SubscriptionRow {
  id: string;
  event_type: string;
  webhook_url: string;
  webhook_headers_json: string | null;
  criteria_channel: string;
  criteria_direction: string | null;
  status: string;
  created_at: string;
  updated_at: string;
}

function toSubscription(row: SubscriptionRow): ZenviaSubscription {
  return {
    id: row.id,
    eventType: row.event_type,
    webhookUrl: row.webhook_url,
    webhookHeaders: row.webhook_headers_json ? JSON.parse(row.webhook_headers_json) : null,
    criteriaChannel: row.criteria_channel,
    criteriaDirection: row.criteria_direction,
    status: row.status,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
}

export function createSubscription(input: CreateSubscriptionInput): ZenviaSubscription {
  const db = getDb();
  const id = randomUUID();
  db.prepare(
    `INSERT INTO zenvia_subscriptions (id, event_type, webhook_url, webhook_headers_json, criteria_channel, criteria_direction, status)
     VALUES (?, ?, ?, ?, ?, ?, 'ACTIVE')`,
  ).run(
    id,
    input.eventType,
    input.webhook.url,
    input.webhook.headers ? JSON.stringify(input.webhook.headers) : null,
    input.criteria.channel,
    input.criteria.direction ?? null,
  );
  const row = db.prepare("SELECT * FROM zenvia_subscriptions WHERE id = ?").get(id) as SubscriptionRow;
  return toSubscription(row);
}

export function listSubscriptions(): ZenviaSubscription[] {
  const rows = getDb().prepare("SELECT * FROM zenvia_subscriptions ORDER BY created_at").all() as SubscriptionRow[];
  return rows.map(toSubscription);
}

export function getSubscription(id: string): ZenviaSubscription | undefined {
  const row = getDb().prepare("SELECT * FROM zenvia_subscriptions WHERE id = ?").get(id) as
    | SubscriptionRow
    | undefined;
  return row ? toSubscription(row) : undefined;
}

export function deleteSubscription(id: string): boolean {
  const result = getDb().prepare("DELETE FROM zenvia_subscriptions WHERE id = ?").run(id);
  return result.changes > 0;
}

export function findActiveSubscriptionsForChannel(channel: string): ZenviaSubscription[] {
  const rows = getDb()
    .prepare(
      `SELECT * FROM zenvia_subscriptions
       WHERE event_type = 'MESSAGE_STATUS' AND status = 'ACTIVE' AND criteria_channel = ?`,
    )
    .all(channel) as SubscriptionRow[];
  return rows.map(toSubscription);
}

File: internal/providers/zenvia/state.go
package zenvia

import (
	"database/sql"
	"encoding/json"
	"errors"

	"github.com/google/uuid"
)

type Message struct {
	ID         string  `json:"id"`
	ExternalID *string `json:"externalId"`
	Channel    string  `json:"channel"`
	Direction  string  `json:"direction"`
	From       string  `json:"from"`
	To         string  `json:"to"`
	Contents   []any   `json:"contents"`
	Status     string  `json:"status"`
	CreatedAt  string  `json:"createdAt"`
}

type Subscription struct {
	ID                string            `json:"id"`
	EventType         string            `json:"eventType"`
	WebhookURL        string            `json:"webhookUrl"`
	WebhookHeaders    map[string]string `json:"webhookHeaders"`
	CriteriaChannel   string            `json:"criteriaChannel"`
	CriteriaDirection *string           `json:"criteriaDirection"`
	Status            string            `json:"status"`
	CreatedAt         string            `json:"createdAt"`
	UpdatedAt         string            `json:"updatedAt"`
}

type Store struct {
	db *sql.DB
}

func NewStore(db *sql.DB) *Store {
	return &Store{db: db}
}

func scanMessage(row interface{ Scan(dest ...any) error }) (*Message, error) {
	var (
		m            Message
		contentsJSON string
	)
	if err := row.Scan(&m.ID, &m.ExternalID, &m.Channel, &m.Direction, &m.From, &m.To, &contentsJSON, &m.Status, &m.CreatedAt); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(contentsJSON), &m.Contents); err != nil {
		return nil, err
	}
	return &m, nil
}

func (s *Store) CreateMessage(channel string, externalID *string, from, to string, contents []any) (*Message, error) {
	id := uuid.NewString()
	contentsJSON, err := json.Marshal(contents)
	if err != nil {
		return nil, err
	}
	_, err = s.db.Exec(
		`INSERT INTO zenvia_messages (id, external_id, channel, direction, sender, recipient, contents_json, status)
		 VALUES (?, ?, ?, 'OUT', ?, ?, ?, 'ACCEPTED')`,
		id, externalID, channel, from, to, string(contentsJSON),
	)
	if err != nil {
		return nil, err
	}
	return s.GetMessage(id)
}

func (s *Store) ListMessages() ([]Message, error) {
	rows, err := s.db.Query(
		`SELECT id, external_id, channel, direction, sender, recipient, contents_json, status, created_at
		 FROM zenvia_messages ORDER BY created_at DESC`,
	)
	if err != nil {
		return nil, err
	}
	defer func() { _ = rows.Close() }()

	out := []Message{}
	for rows.Next() {
		m, err := scanMessage(rows)
		if err != nil {
			return nil, err
		}
		out = append(out, *m)
	}
	return out, rows.Err()
}

func (s *Store) GetMessage(id string) (*Message, error) {
	row := s.db.QueryRow(
		`SELECT id, external_id, channel, direction, sender, recipient, contents_json, status, created_at
		 FROM zenvia_messages WHERE id = ?`, id,
	)
	m, err := scanMessage(row)
	if errors.Is(err, sql.ErrNoRows) {
		return nil, nil
	}
	return m, err
}

func (s *Store) UpdateMessageStatus(id, status string) error {
	_, err := s.db.Exec("UPDATE zenvia_messages SET status = ? WHERE id = ?", status, id)
	return err
}

func scanSubscription(row interface{ Scan(dest ...any) error }) (*Subscription, error) {
	var (
		sub         Subscription
		headersJSON sql.NullString
		criteriaDir sql.NullString
	)
	if err := row.Scan(&sub.ID, &sub.EventType, &sub.WebhookURL, &headersJSON, &sub.CriteriaChannel, &criteriaDir, &sub.Status, &sub.CreatedAt, &sub.UpdatedAt); err != nil {
		return nil, err
	}
	if headersJSON.Valid {
		if err := json.Unmarshal([]byte(headersJSON.String), &sub.WebhookHeaders); err != nil {
			return nil, err
		}
	}
	if criteriaDir.Valid {
		sub.CriteriaDirection = &criteriaDir.String
	}
	return &sub, nil
}

type CreateSubscriptionInput struct {
	EventType         string
	WebhookURL        string
	WebhookHeaders    map[string]string
	CriteriaChannel   string
	CriteriaDirection *string
}

func (s *Store) CreateSubscription(input CreateSubscriptionInput) (*Subscription, error) {
	id := uuid.NewString()
	var headersJSON *string
	if input.WebhookHeaders != nil {
		b, err := json.Marshal(input.WebhookHeaders)
		if err != nil {
			return nil, err
		}
		str := string(b)
		headersJSON = &str
	}
	_, err := s.db.Exec(
		`INSERT INTO zenvia_subscriptions (id, event_type, webhook_url, webhook_headers_json, criteria_channel, criteria_direction, status)
		 VALUES (?, ?, ?, ?, ?, ?, 'ACTIVE')`,
		id, input.EventType, input.WebhookURL, headersJSON, input.CriteriaChannel, input.CriteriaDirection,
	)
	if err != nil {
		return nil, err
	}
	return s.GetSubscription(id)
}

func (s *Store) ListSubscriptions() ([]Subscription, error) {
	rows, err := s.db.Query(
		`SELECT id, event_type, webhook_url, webhook_headers_json, criteria_channel, criteria_direction, status, created_at, updated_at
		 FROM zenvia_subscriptions ORDER BY created_at`,
	)
	if err != nil {
		return nil, err
	}
	defer func() { _ = rows.Close() }()

	out := []Subscription{}
	for rows.Next() {
		sub, err := scanSubscription(rows)
		if err != nil {
			return nil, err
		}
		out = append(out, *sub)
	}
	return out, rows.Err()
}

func (s *Store) GetSubscription(id string) (*Subscription, error) {
	row := s.db.QueryRow(
		`SELECT id, event_type, webhook_url, webhook_headers_json, criteria_channel, criteria_direction, status, created_at, updated_at
		 FROM zenvia_subscriptions WHERE id = ?`, id,
	)
	sub, err := scanSubscription(row)
	if errors.Is(err, sql.ErrNoRows) {
		return nil, nil
	}
	return sub, err
}

func (s *Store) DeleteSubscription(id string) (bool, error) {
	res, err := s.db.Exec("DELETE FROM zenvia_subscriptions WHERE id = ?", id)
	if err != nil {
		return false, err
	}
	n, err := res.RowsAffected()
	return n > 0, err
}

func (s *Store) FindActiveSubscriptionsForChannel(channel string) ([]Subscription, error) {
	rows, err := s.db.Query(
		`SELECT id, event_type, webhook_url, webhook_headers_json, criteria_channel, criteria_direction, status, created_at, updated_at
		 FROM zenvia_subscriptions
		 WHERE event_type = 'MESSAGE_STATUS' AND status = 'ACTIVE' AND criteria_channel = ?`,
		channel,
	)
	if err != nil {
		return nil, err
	}
	defer func() { _ = rows.Close() }()

	out := []Subscription{}
	for rows.Next() {
		sub, err := scanSubscription(rows)
		if err != nil {
			return nil, err
		}
		out = append(out, *sub)
	}
	return out, rows.Err()
}
